### PR TITLE
refactor: use `Readonly` types in builders where applicable

### PR DESCRIPTION
**Pull request by suneettipirneni**
**Please describe the changes this PR makes and why it should be merged:**

Changes return types and some parameter types to use readonly inputs/outputs.

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating

### CODE REVIEW

**Comment by almeidx** [comment]
Would `readonly T[]` syntax be preferred over `ReadonlyArray<T>` given that `T[]` is already being used instead of `Array<T>`?

**Comment by suneettipirneni** [comment]
> Would `readonly T[]` syntax be preferred over `ReadonlyArray<T>` given that `T[]` is already being used instead of `Array<T>`?

Pretty sure `readonly T[]` isn't a valid syntax.

**Comment by almeidx** [comment]
> Pretty sure `readonly T[]` isn't a valid syntax.

It is ([handbook](https://www.typescriptlang.org/docs/handbook/2/objects.html#the-readonlyarray-type))

**Comment by suneettipirneni** [comment]
> > Pretty sure `readonly T[]` isn't a valid syntax.
> 
> It is ([handbook](https://www.typescriptlang.org/docs/handbook/2/objects.html#:~:text=Just%20as%20TypeScript%20provides%20a%20shorthand%20syntax%20for%20Array%3CType%3E%20with%20Type%5B%5D%2C%20it%20also%20provides%20a%20shorthand%20syntax%20for%20ReadonlyArray%3CType%3E%20with%20readonly%20Type%5B%5D.))

Ah ok I'll adjust them then.

**Review by iCrawl — APPROVED** [review]

File: packages/builders/src/components/ActionRow.ts [hunk 1]
```diff
@@ -27,7 +27,7 @@ export class ActionRow<T extends ActionRowComponent = ActionRowComponent> extend
 	 * @param components The components to add to this action row.
 	 * @returns
 	 */
-	public addComponents(...components: T[]) {
+	public addComponents(...components: ReadonlyArray<Readonly<T>>) {
```
Comment (kyranet) [inline packages/builders/src/components/ActionRow.ts:30]:
```suggestion
	public addComponents(...components: readonly Readonly<T>[]) {
```

File: packages/builders/src/messages/embed/Assertions.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ export const embedFieldsArrayPredicate = embedFieldPredicate.array();
 
 export const fieldLengthPredicate = z.number().lte(25);
 
-export function validateFieldLength(amountAdding: number, fields?: APIEmbedField[]): void {
+export function validateFieldLength(amountAdding: number, fields?: ReadonlyArray<Readonly<APIEmbedField>>): void {
```
Comment (kyranet) [inline packages/builders/src/messages/embed/Assertions.ts:20]:
```suggestion
export function validateFieldLength(amountAdding: number, fields?: readonly Readonly<APIEmbedField>[]): void {
```

File: packages/builders/src/messages/embed/UnsafeEmbed.ts [hunk 3]
```diff
@@ -299,7 +300,7 @@ export class UnsafeEmbed implements Equatable<APIEmbed | UnsafeEmbed> {
 	 *
 	 * @param timestamp The timestamp or date
 	 */
-	public setTimestamp(timestamp: number | Date | null = Date.now()): this {
+	public setTimestamp(timestamp: number | Readonly<Date> | null = Date.now()): this {
```
Comment (kyranet) [inline packages/builders/src/messages/embed/UnsafeEmbed.ts:303]: We really need [ReadonlyDate](https://github.com/microsoft/TypeScript/issues/29274), heh.

`Readonly<Date>` is a bit useless here, it doesn't have public fields, only methods.

**Review by kyranet — CHANGES_REQUESTED** [review]

**Comment by BenPatersonxx** [comment]
@suneettipirneni

**Pull request closed without merging**